commitlog: Keep track of high RP for flush request on a per table basis

elcallio opened the pull request:
Refs #23656
Fixes #28421

Keep track of each table high RP individually, instead of using the max range of all segments being flush-requested. The idea is to allow some tables to ignore a subsequent flush, close after an earlier one, if no additional data has been added that falls under the range being fr:d.

Instead of keeping track of "high" flush position, keep track of lowest written
replay position per compaction group.
    
This helps ensures we don't hit aliasing issues when trying to filter out memtables from subsequent flush request, while previous flush is still pending.

Since a flush request from commit log will always be on segment boundary, unless we track exact low position in table/compaction group, we can get false overlap between memtable generations.

Moving the counter to compaction group also makes it possible to only "partially" flush a table, if different compaction group ranges have different CL segment usage (very rare, but...).

Note: since we only keep track of the "low" counter here, removing/adding/replacing a compaction group will not affect truncation or other bookkeep that relies on the table "high" counters.

Commit:
commitlog: Keep track of high RP for flush request on a per table basis

Refs #23656

Keep track of each table high RP individually, instead of using the max
range of all segments being flush-requested. The idea is to allow some
tables to ignore a subsequent flush, close after an earlier one, if no
additional data has been added that falls under the range being fr:d.

Commit:
commitlog: Keep track of high RP for flush request on a per table basis

Refs #23656

Keep track of each table high RP individually, instead of using the max
range of all segments being flush-requested. The idea is to allow some
tables to ignore a subsequent flush, close after an earlier one, if no
additional data has been added that falls under the range being fr:d.

elcallio commented:
Note. This change does not make anything work magically different. 

Flush requests flow is basically:

- table A: holds data in RP range 0 -> N

- cl requests flush of range 0 -> M (M < N). A will now flush _all_ data, i.e. get a position N for "last flush pos". 

- cl requests flush of range M -> M'  (M' > M, M' < N). A can now skip flush, since last flush pos > M'. 

This patch only helps in cases where data in CL segments is distributed such that table A could reduce a range at a flush that might otherwise cover range higher than its flush pos, but data is only located in segments lower than this, in which case we reduce the request range, and might bypass a flush until a later request.

Testing on the repro benchmark @avikivity wrote, I get (it seems) slightly better overall result, in that the overall distribution of small sstables is lower, and converges (via compaction) slightly faster. Not conclusive though. But it does suggest that in a similar situation there might be occurrences where we can indeed bypass a table flush sometimes.

avikivity commented:
I think what's necessary is to consult the memtables' replay position (I see we already keep it).

For each memtable,

 - If the flush replay position is higher than a memtable's replay position, ignore that memtable
 - Otherwise,
     - If the memtable is active, seal it and wait for it to flush,
     - Otherwise, wait for the ongoing flush to complete

elcallio commented:
> I think what's necessary is to consult the memtables' replay position (I see we already keep it).
> 
> For each memtable,
> 
>     * If the flush replay position is higher than a memtable's replay position, ignore that memtable
> 
That is not what the respective RP:s represent.

* table highest_rp is, unsurprisingly, the highest RP of mutations added to table
* table flush_pos is at which point we last did a flush, i.e. highest pos flushed
* CL flush calls with a position which is the highest position we flush. Implicitly, this is the range 0->pos (or rather, last->pos, but they mean the same, since we request flush in order, and sequentially). 

We can ignore a table (memtable) if its _flush_ position is _higher_ than the position we request flush to. 
Calling with a value higher or equal to table flush_pos means we request a range covered by the memtable data. 

Again, since all these are sequential ranges, there are very few cases where we _don't_ overlap the requested range. 

Technically, we could add a low position to table/compaction group, and potentially filter out memtables on "full" overlap semantics, but again, the sequential nature here means that the chance of this really adding anything is pretty low; We are almost guaranteed to have a low RP lower or equal to the previous flush request position, unless there was somehow a write pause that lasted one or more segments (possible, but imho unlikely in a real world, high throughput situation). 
Still, maybe worth experimenting with...

elcallio commented:
No, it can be done. What I am saying is that while the current mechanism is a little coarse in the filtering, and this patch makes it a little more narrow, the nature of how mutations are written to and later discarded from the commit log means that for most cases memtables _will_ be overlapping the segment(s) being released in some form or other. 
We can try to narrow the filter even further, but I fear it will still be the case, unless writes are (very) rare, that regardless, most memtables will overlap. 

To add some more clarification:

Say we fill the commitlog to the halfway mark, and now start trying to request flushes; We will try to get rid of as many bytes as we are over the flush threshold, one or more segments, representing the oldest (i.e. lowest RP range) mutations added so far. 

We find a range 0->N, and request flushing for each table having data (remaining) in said segments. These will each have one active memtable per compaction group. Assuming general pressure or anything else did not cause flushing memtables to sstable previously, each of these memtables will have overlapping data. 
Since we cannot flush "half" a memtable, the full range for each table will be written to disk.
So for a table A, we will write data representing range 0->X, where X >= N. 

The optimization we can do here (and which is improved by this patch) is that if we, next time we cross the CL threshold, request flush now for range N->S, S <= X, table A can ignore this request, as we've already handled it. 

The same applies of course if memtable flushes are done organically, and the filtering can apply here as well.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16683/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16683/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4691/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7857/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15906/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16683/)
#### Build Details:
- Duration: 6 hr 58 min
- Builder: spider7.cloudius-systems.com

avikivity commented:
> No, it can be done. What I am saying is that while the current mechanism is a little coarse in the filtering, and this patch makes it a little more narrow, the nature of how mutations are written to and later discarded from the commit log means that for most cases memtables _will_ be overlapping the segment(s) being released in some form or other. We can try to narrow the filter even further, but I fear it will still be the case, unless writes are (very) rare, that regardless, most memtables will overlap.
> 
> To add some more clarification:
> 
> Say we fill the commitlog to the halfway mark, and now start trying to request flushes; We will try to get rid of as many bytes as we are over the flush threshold, one or more segments, representing the oldest (i.e. lowest RP range) mutations added so far.
> 
> We find a range 0->N, and request flushing for each table having data (remaining) in said segments. These will each have one active memtable per compaction group. Assuming general pressure or anything else did not cause flushing memtables to sstable previously, each of these memtables will have overlapping data. Since we cannot flush "half" a memtable, the full range for each table will be written to disk. So for a table A, we will write data representing range 0->X, where X >= N.
> 
> The optimization we can do here (and which is improved by this patch) is that if we, next time we cross the CL threshold, request flush now for range N->S, S <= X, table A can ignore this request, as we've already handled it.
> 
> The same applies of course if memtable flushes are done organically, and the filtering can apply here as well.

I don't see why. Let's assume an overwrite workload (so no memory-pressure flush). We write to a memtable, commitlog reaches the halfway mark, we seal and flush the memtable and open a new one.

The seal-and-flush affects all commitlog segments up to the youngest one (call it Y). So when the next segment reaches the halfway mark, it will not have anything to flush, and so forth, until Y reaches the halfway mark. When we flush the memtables with footprint in Y, again we will free the entire commitlog.

elcallio commented:
> I don't see why. Let's assume an overwrite workload (so no memory-pressure flush). We write to a memtable, commitlog reaches the halfway mark, we seal and flush the memtable and open a new one.
> 
> The seal-and-flush affects all commitlog segments up to the youngest one (call it Y). So when the next segment reaches the halfway mark, it will not have anything to flush, and so forth, until Y reaches the halfway mark. When we flush the memtables with footprint in Y, again we will free the entire commitlog.

I don't understand what you are trying to say. Half-way mark flush threshold is per the whole commitlog (total disk usage). 
We flush as much as we need to get below the mark again; This will indeed cause the full memtable here, and record a flush position equal to the highest RP it had recorded. In this scenario that would be a position not touched by the flush request, probably in the active segment.

Note then that this memtable flush will release all references to the segment it held data relevant for. This means that either, if this is the only table in the system, _all_ segments were declared clean suddenly, and our footprint shrank much more than we hoped for, _or_, if there are other tables, all other segments at least stopped referencing this table (and its memtables), so we will not really be called on any further flush requests anyway.

elcallio commented:
Fwiw, something happened in master/seastar during today/yesterday (_or_ my machine is in a different mode). After rebase/pull, I get much better numbers for the #23656 benchmark, (master, at least with tablets enabled), and much better convergence towards balance/reasonable sstable numbers. Of course, YMMW.

Also noting; This PR has much better convergence, and slightly better number in vnode mode, than master (for me), suggesting that the pattern of "organic flush partially overlapping CL flush + segments with not all tables represented" might be more prevalent there.

Rebased this.

scylladb-promoter commented:
#### :red_circle: CI State: UNSTABLE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16834/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16834/)
:x: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8003/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16071/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4830/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16834/)
#### Build Details:
- Duration: 6 hr 39 min
- Builder: spider6.cloudius-systems.com

avikivity commented:
My point wasn't about what's happening during the flushing of the first segment to hit the halfway mark (call it N), but the one after it (call it N+1).

N flushed memtable M. If this flush causes M's link with N+1 to be unlinked, we're good.

If it doesn't (perhaps M's flush is delayed), then when N+1 reaches the halfway mark, it will also cause the table to be flushed, this time flushing M+1 (which will be tiny).

elcallio commented:
For the most part, no. Because we place the flush position mark in table, flushing requests for remaining segments will not cause any flushes of data not pertaining to them.

More ascii art:

```
Time 0:

CL:  [seg    1][seg     2][seg     3]....[seg     N]

T1:          [MT A, RP = seq 3'] 
```
Flush request for "seg 1" will cause T1's memtable to be flushed, setting "flush position to "seq 3'" (a range inside seq 3).

```
Time 1:
CL:  ---------[seg     2][seg     3]....[seg     N]

T1:          -------------------[MT B, RP = seq N-M'] 

```
If we now flush request "seg 2", T1's new memtable will _not_ be flushed, since T1 flush position is already  higher than the "seq 2" mark. 
Only once we actually hit "seg 3" will T1 flush again, since we now again overlap the RP range.

Commit:
replay_position: Add a "max" constant

Commit:
replay_position: Add a "max" constant

Commit:
table/compaction group: Move flush replay_position guard into CG "lowest" pos

Refs #23656

Instead of keeping track of "high" flush position, keep track of lowest written
replay position per compaction group.

This helps ensures we don't hit aliasing issues when trying to filter out memtables
from subsequent flush request, while previous flush is still pending.

Since a flush request from commit log will always be on segment boundary,
unless we track exact low position in table/compaction group, we can get false
overlap between memtable generations.

Moving the counter to compaction group also makes it possible to only "partially"
flush a table, if different compaction group ranges have different CL segment
usage (very rare, but...).

Note: since we only keep track of the "low" counter here, removing/adding/replacing
a compaction group will not affect truncation or other bookkeep that relies
on the table "high" counters.

v2:
* Move reset of low rp mark to actual flush routine, to ensure we
  reset the value on organic (ditry mem manager initiated) flush
  as well as CL initiated.

Commit:
table/compaction group: Move flush replay_position guard into CG "lowest" pos

Refs #23656

Instead of keeping track of "high" flush position, keep track of lowest written
replay position per compaction group.

This helps ensures we don't hit aliasing issues when trying to filter out memtables
from subsequent flush request, while previous flush is still pending.

Since a flush request from commit log will always be on segment boundary,
unless we track exact low position in table/compaction group, we can get false
overlap between memtable generations.

Moving the counter to compaction group also makes it possible to only "partially"
flush a table, if different compaction group ranges have different CL segment
usage (very rare, but...).

Note: since we only keep track of the "low" counter here, removing/adding/replacing
a compaction group will not affect truncation or other bookkeep that relies
on the table "high" counters.

v2:
* Move reset of low rp mark to actual flush routine, to ensure we
  reset the value on organic (ditry mem manager initiated) flush
  as well as CL initiated.

elcallio commented:
Updated this PR: 

* Upgraded from RFC. 
* Added patch to make sure flush range granularity/aliasing is minimised by changing the flush position into an actual, per-compaction group, low position. This adds a variable update to every write, but should make the range exclusion on CL requested flush be exact (not alias at segment boundaries).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16913/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16913/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4887/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8058/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16128/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16913/)
#### Build Details:
- Duration: 6 hr 34 min
- Builder: spider5.cloudius-systems.com

bhalevy commented:
@elcallio is this PR ready for merge?
Cc @regevran.
Please push forward to completion.

bhalevy commented:
> It is mergable (afaik). Note that this change will add another RP write for each mutation apply, i.e. a very slight overhead (cmp + storing a member variable). It is possible to make roughly the same granularity tightening by extrapolating this value at flush instead. It will be slightly less accurate for organically flushed memtables, but would save the cmp+store on every write.

@elcallio less overhead would be better

elcallio commented:
Ran the bench from #23636 with this vs master, looking at memtable occupancy when flushing. At 32MB segment size, there is _maybe_ a slight edge in this PR. At 8MB segments, it is possibly higher, but we are not talking breathtaking differences. 
Fwiw.

elcallio commented:
The more I think about it, the less useful running tests seem. You can summarise the problem quite easily;

Commitlog represents the distribution of data in the system over time (where time is defined as write order per shard). Its life cycle is basically a circular buffer from which we want to shed in the "old" end. 

Tablets represent the distribution of data over key topology/distribution. Only one single memtable exists "living" at any time, which is essentially an area of time and key range.

Now, given any benchmark or even real world application, the distribution of keys written over time is very likely to be approximating a normal distribution. The rate at which we write will determine how much of the key range space we fill for any time range. 

The flushing represents us discarding a time slice of data. The probability that a tablet/compaction group/memtable will intersect this time slice will be directly proportional to the write rate, the time slice size (i.e. segment size) and the frequency with which we require flushing, which in turn is proportional to total size and write rate. 

I don't think we can make the distinction between intersections of the two axis above much more precise that is done here (suggestions welcome). Unless, however, "natural" flushing has an equal or higher rate that total influx of writes (the fact that deletes etc are included does not help), the odds of any given tablet having intersection with a time slice being flushed will remain pretty high. 

Shrinking segments increase the probability of non-overlap, but that has a whole other can of worms attached (write efficiency). Increasing the overall size as suggested might help quite a bit, esp. if you over provision such that the CL size is twice the memory size or similar. 

I must come back to again; What is the fictional/real workload we'd like to optimize here?

avikivity commented:
Either this is helping something, or it is not. If it is helping something, it should be demonstrable with a (manual) test.

If it's not helping something, then we shouldn't commit it.

To me it seems obvious that narrowing the scope that a commitlog segment memtable flush is helpful, because it lets memtables that escape the narrowed scope grow larger. Larger memtables = lower write amplification.

The question is whether the narrowing is sufficient. Maybe it's not. If it is, we should be able to run a test that demonstrates it.

Why not go to the narrowest possible scope, a memtable? If a memtable has some footprint in a commitlog segment, then flush it when the segment hits the soft flush threshold, and wait for the flush when it hits the hard limit.

The goal is to avoid flushing memtables that were born recently and have no intersection with the commitlog segment that we're flushing.

tgrabiec commented:
@elcallio Did you manage to write a reproducer which demonstrates the improvement?

bhalevy commented:
> @elcallio Did you manage to write a reproducer which demonstrates the improvement?

ping @elcallio

elcallio commented:
> @elcallio Did you manage to write a reproducer which demonstrates the improvement?

As stated above, I can see some lessened aliasing with this applied, if segment size/write rate etc are at specific levels. Did I write a repro for which this solves any significant issues? No.

denesb commented:
> Did I write a repro for which this solves any significant issues? No.

You didn't write because it is not possible / too artificial / not relevant or because you didn't get around to do it?

avikivity commented:
> We need a decision from @avikivity and @tgrabiec if they accept the PR in its present form. AFAU @elcallio reached a dead-end trying to come up with a reproducer which would show significant benefits. So the PR is a minor improvement (with no real drawback either).

If we can't figure out where it helps, maybe it doesn't help.

bhalevy commented:
@mykaul this is an important fix that should be considered as a release blocker for 2026.1 IMO.
What do you think?

avikivity commented:
> @elcallio #23656 was closed. What's the issue that this PR fixes / refs. Also, why shouldn't it be backported?

Nothing should be backported without a positive reason.

avikivity commented:
> The issue is #26823
> 
> > This does not fix a regression.
> 
> It is a regression compared with vnodes. I believe that we suffered from this in the field when a customer upgraded to 2025.3, which increased tablet count per shard from 1 to 10 (now default), which increased memtable flush rate and compaction load, and resulted in elevated request latency. The problem is more pronounced with a higher tablet count per shard.
> 
> If commitlog segments are 32MB, with 8GB shards we can expect 200 segments, so memtables will be 200 times smaller due to this.

Then I agree it needs to be backported. But it needs to be explained clearly. And demonstrated! 

btw memtables are expected to be 100X smaller (if we have one table and 100 tablets), but not 200X smaller.

avikivity commented:
> I verified that [small memtable flushing phenomena](https://github.com/scylladb/scylladb/pull/23678#issuecomment-3792581915) doesn't happen with this PR, and all memtables have reasonable flush size.
> 
> @avikivity Do you want additional testing before merging?

The patch changelog wants a test scenario (machine configuration and workload) and results (say, memtable size distribution at flush time) before and after the patch.

mykaul commented:
> We need soak time to make sure it doesn't cause other problems.
> 
> I'm not at all happy with our commitlog tests (the unit tests are fine, I'm more worried about the systematic test to show that we never free a commitlog segment unless related memtables are flushed), so the least we can do is give it some time.

@avikivity - is it time to backport?
